synapse/rest/client/_base.py

```diff
@@ -76,7 +76,9 @@ def set_timeline_upper_limit(filter_json: JsonDict, filter_timeline_limit: int)
         )
 
 
-def interactive_auth_handler(orig):
+def interactive_auth_handler(
+    orig: Callable[..., Awaitable[Tuple[int, JsonDict]]]
+) -> Callable[..., Awaitable[Tuple[int, JsonDict]]]:
```

Comment:
Since we're here we may as well factor out a `C = TypeVar("C", bound=Callable[..., Awaitable[Tuple[int, JsonDict]]])`? Otherwise using `@interactive_auth_handler` will strip out a bunch of the typing?

Reply:
Maybe. I haven't been able to figure out what using the `bound` keyword actually does for typevars, but I'll take another look.

Reply:
I tried this (and changed `def interactive_auth_handler(orgi: C) -> C`), but it then gave an error about the return value not equaling the expected value.

Reply:
> Maybe. I haven't been able to figure out what using the `bound` keyword actually does for typevars, but I'll take another look.

The bound basically means that you can only use the `@interactive_auth_handler` on things that are `Callable[..., Awaitable[Tuple[int, JsonDict]]]`



> I tried this (and changed `def interactive_auth_handler(orgi: C) -> C`), but it then gave an error about the return value not equaling the expected value.

Yeah, annoyingly you need to add a `cast(C, warpped)` when returning, an example of this can be found at e.g. https://github.com/matrix-org/synapse/blob/e2481dbe9325321d460037a2efe9b9ea2ac78057/synapse/util/caches/lrucache.py#L407-L413